keps/sig-multicluster/1645-multi-cluster-services-api/README.md

```diff
@@ -0,0 +1,925 @@
+ServiceExport CR:
+
+```golang
+// ServiceExport declares that the associated service should be exported to
+// other clusters.
+type ServiceExport struct {
+        metav1.TypeMeta `json:",inline"`
+        // +optional
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        // +optional
+        Status ServiceExportStatus `json:"status,omitempty"`
+}
+
+// ServiceExportStatus contains the current status of an export.
+type ServiceExportStatus struct {
+        // +optional
+        // +patchStrategy=merge
+        // +patchMergeKey=type
+        // +listType=map
+        // +listMapKey=type
+        Conditions []ServiceExportCondition `json:"conditions,omitempty"`
+}
+
+// ServiceExportConditionType identifies a specific condition.
+type ServiceExportConditionType string
+
+const {
+      // ServiceExportInitialized means the service export has been noticed
+      // by the controller, has passed validation, has appropriate finalizers
+      // set, and any required supercluster resources like the IP have been
+      // reserved
+      ServiceExportInitialized ServiceExportConditionType = "Initialized"
+      // ServiceExportExported means that the service referenced by this
+      // service export has been synced to all clusters in the supercluster
+      ServiceExportExported ServiceExportConditionType = "Exported"
+}
+
+// ServiceExportCondition contains details for the current condition of this
+// service export.
+//
+// Once [#1624](https://github.com/kubernetes/enhancements/pull/1624) is
+// merged, this will be replaced by metav1.Condition.
+type ServiceExportCondition struct {
+        Type ServiceExportConditionType `json:"type"`
+        // Status is one of {"True", "False", "Unknown"}
+        Status corev1.ConditionStatus `json:"status"`
+        // +optional
+        LastTransitionTime *metav1.Time `json:"lastTransitionTime,omitempty"`
+        // +optional
+        Reason *string `json:"reason,omitempty"`
+        // +optional
+        Message *string `json:"message,omitempty"`
+}
+```
+```yaml
+apiVersion: multicluster.k8s.io/v1alpha1
+kind: ServiceExport
+metadata:
+  name: my-svc
+  namespace: my-ns
+status:
+  conditions:
+  - type: Initialized
+    status: "True"
+    lastTransitionTime: "2020-03-30T01:33:51Z"
+  - type: Exported
+    status: "True"
+    lastTransitionTime: "2020-03-30T01:33:55Z"
+```
+
+`ServiceExports` will be created within the cluster and namespace that the
+service resides in and are name-mapped to the service for export - that is, they
+reference the `Service` with the same name as the export. If multiple clusters
+within the supercluster have `ServiceExports` with the same name and namespace,
+these will be considered the same service and will be combined at the
+supercluster level.
+
+This requires that within a supercluster, a given namespace is governed by a
+single authority across all clusters. It is that authority’s responsibility to
```

Comment:
Is there an expectation that a "normal" namespace admin has the authority to export his service across the super cluster?  Is there a plan that would allow a cluster-admin to restrict which namespaces may have serviceexport and service import?